.github/workflows/meson.yml

```diff
@@ -67,25 +70,27 @@ jobs:
         if: matrix.config.os == 'windows-latest' && matrix.config.msvc == true
         uses: ilammy/msvc-dev-cmd@v1
 
-      - name: Install MSVC dependecies with choco
+      - name: Get time
         if: matrix.config.msvc
+        id: get_time
         run: |
-          # for whatever reason choco insist on a reboot if it installs meson
-          choco install ninja nasm python pkgconfiglite
-          python -m pip install --upgrade pip
-          pip install meson
+          "timestamp=$((Get-Date).Ticks)" >> $env:GITHUB_OUTPUT
 
-      - name: Install more MSVC dependecies with msys
+      - name: Restore cache
+        if: matrix.config.msvc
+        uses: actions/cache/restore@v4
+        with:
+          path: |
+            ${{ env.SCCACHE_DIR }}
+            ${{ github.workspace }}/subprojects
```

Comment:
Does `meson wrap update-db`  update existing project downloads?
While the meson PR was being wroked on we had some wrap files directly in the repo and back then meson just always kept existing project sources around leading to issues everytime a dep was upgraded.

Reply:
No, it does not update. wrap db is just a json file with list of available projects, but once they are initialized in subproject directory they will not be updated until manually changed. Things in subproject directory always takes precedence.

> meson just always kept existing project sources around leading to issues everytime a dep was upgraded.

Yes, this is the case. As far as I know, the only solution is to clear subprojects driectory manually.

Reply:
If there’s a straightforward way of retrieving the availbe version on wrapdb, they could be included in the cache key. Otherwise you could limit the lifetime of the subproject download cache like done here to avoid us getting stuck on the same dep versions forever: https://github.com/libass/libass-tests/blob/cee2df5fa840df02dda1a33f75a99844b5cfd61e/.github/workflows/art.yml#L124

(Note: the linked setup also includes an additional `rev` number to allow easily forcing a cache rebuild if necessary, but since GHA now allows deleting caches from the web UI this shouldn't be necessary anymore)

Reply:
After taking a look into it, `--reset` will remove stale sources, so all we have to do is `meson subprojects update --reset`.

Reply:
hmm.. it’s not documented at [the usual place]() and the code looks like it indiscriminately deletes even up-to-date subdirs and but reuse already downloaded tarballs. Those tarball names are versioned, so that’s fine. However:
- the cache will slowly grow since old tarballs are never cleaned up
- `meson subprojects update` doesn't recognize extracted project dirs anymore if the wrap file version changed (since the versioned dir name changed); so they too never get cleaned up:

    ```
    Updating fribidi-1.0.15...
      -> Cannot update subproject with no wrap file
    Using fribidi source from cache.
    Updating fribidi...
      -> New version extracted
    ```
  
- neither `update` nor `wrap update-db` actually seem to modify an existing `*.wrap` file. To upgrade fribidi i had to first delete the existing wrap

How did you get this working on your end?

Reply:
> hmm.. it’s not documented at [the usual place](https://github.com/libass/libass/pull/857) and the code looks like it indiscriminately deletes even up-to-date subdirs and but reuse already downloaded tarballs.

I think your link is not correct, but it is documented here https://mesonbuild.com/Subprojects.html#update-subprojects And yes, what it does it to remove source dir and update it from .wrap file, which would otherwise never done.

> the cache will slowly grow since old tarballs are never cleaned up

Typical cache issue, sccache will also grow with old object files that are not used. Up to limit.

I don't think the amount and frequency of updates is that big to warrant added complexity to manage that. Even if it grows with a "few" of older version at some point, it can be purged manually, from GHA cache.

> doesn't recognize extracted project dirs anymore if the wrap file version changed (since the versioned dir name changed); so they too never get cleaned up:

Yes, indeed. And I tried the [`purge`](https://mesonbuild.com/Release-notes-for-0-58-0.html#purge-subprojects-folder) option seems to remove new version instead of old one. Things are strange. 

(Also it is really annoying that Meson document functions in release notes only and don't put it in the normal documentation...)

Anyway, I changed it now to cache only `subprojects/packagecache`, so tarballs and patches. So we avoid any wrap/source dir issues, because it clearly is not designed to be managed in automatic way.

> neither update nor wrap update-db actually seem to modify an existing *.wrap file. To upgrade fribidi i had to first delete the existing wrap

To update wrap, you can use `meson wrap update`, which is NOT the same as `meson subproject update`. But not important anymore, wrap will be recreated each time and only cache will be reused.

Reply:
> I think your link is not correct, 

Yeah, I forgot to actually copy the link and GH interprets an empty `[link text]()` as a link to the current page, sorry.
This is the link I meant to reference *(`purge` is documented, but `--reset` is absent; annoying indeed)*: https://mesonbuild.com/Commands.html#subprojects

> Typical cache issue, sccache will also grow with old object files that are not used. Up to limit.

**“Up to limit”** being the important distinction here. For sccache *(and other sane caches)* we basically allocate a chunk of our cahce budget to this (the max limit)*, but as implemented the subprojects cache will grow indefinitely. It growing slowly just means it’s less likely someone keeps an eye on it or immediately remembers what to do once problems show up, or jobs slow down from down/uploading bloated cache entries.

Given we have a known set of project names to worry about, properly managing amounts to just grouping all file entries matching up to their first digit, sorting the tail part starting from the first digit numerically and deleting all but the (something like: `sort -nr | tail -n +2 | xargs -n 1 printf '%s%s\n' "$project" | xargs rm`)

>

Reply:
I've added a cache clearing step. I still think it is not strictly necessary though. For two reasons. First even few years worth of tarballs for our dependencies would produce any real issue with the cache itself. And second and more importantly GitHub cache lifetime is 7 days, so unless libass will keep the cache up for literally years by bumping it every few days, it would be naturally cleaned. But sure there is theoretical possibility that it will grow...